File: sudoku/sudoku_v1.js
function sd_genmat() {
	var i, j, r, c, c2, C = [], R = []
	for (i = r = 0; i < 9; ++i)
		for (j = 0; j < 9; ++j)
			for (k = 0; k < 9; ++k)
				C[r++] = [ 9 * i + j, (Math.floor(i/3)*3 + Math.floor(j/3)) * 9 + k + 81, 9 * i + k + 162, 9 * j + k + 243 ]
	for (c = 0; c < 324; ++c) R[c] = []
	for (r = 0; r < 729; ++r)
		for (c2 = 0; c2 < 4; ++c2)
			R[C[r][c2]].push(r)
	return [R, C];
}

function sd_update(R, C, sr, sc, r, v) {
	for (var c2 = 0; c2 < 4; ++c2) {
		var c = C[r][c2];
		sc[c] += v;
		for (var r2 = 0; r2 < 9; ++r2) sr[R[c][r2]] += v;
	}
}

function sd_solve(R, C, _s) {
	var i, j, r, c, r2, dir, c0, hints = 0;
	var sr = [], sc = [], cr = [], cc = [], out = [], ret = [];
	for (r = 0; r < 729; ++r) sr[r] = 0;
	for (c = 0; c < 324; ++c) sc[c] = 0;
	for (i = 0; i < 81; ++i) {
		var a = _s.charAt(i) >= '1' && _s.charAt(i) <= '9'? _s.charCodeAt(i) - 49 : -1;
		if (a >= 0) sd_update(R, C, sr, sc, i * 9 + a, 1);
		if (a >= 0) ++hints;
		cr[i] = cc[i] = -1, out[i] = a + 1;
	}
	for (i = c0 = 0, dir = 1;;) {
		while (i >= 0 && i < 81 - hints) {
			if (dir == 1) {
				var min = 10, n;
				for (var j = 0; j < 324; ++j) {
					var p;
					c = j + c0 < 324? j + c0 : j + c0 - 324
					if (sc[c]) continue;
					for (r2 = n = 0, p = R[c]; r2 < 9; ++r2)
						if (sr[p[r2]] == 0) ++n;
					if (n < min) min = n, cc[i] = c, c0 = c + 1;
					if (min <= 1) break;
				}
				if (min == 0 || min == 10) cr[i--] = dir = -1;
			}
			c = cc[i];
			if (dir == -1 && cr[i] >= 0) sd_update(R, C, sr, sc, R[c][cr[i]], -1);
			for (r2 = cr[i] + 1; r2 < 9; ++r2)
				if (sr[R[c][r2]] == 0) break;
			if (r2 < 9) {
				sd_update(R, C, sr, sc, R[c][r2], 1);
				cr[i++] = r2; dir = 1;
			} else cr[i--] = dir = -1;
		}
		if (i < 0) break;
		var y = []
		for (j = 0; j < 81; ++j) y[j] = out[j]
		for (j = 0; j < i; ++j) r = R[cc[j]][cr[j]], y[Math.floor(r/9)] = r%9 + 1;
		ret.push(y)
		--i; dir = -1;
	}
	return ret;
}

var l, e = sd_genmat()
while ((l = readline()) != null) {
	if (l.length >= 81) {
		var r = sd_solve(e[0], e[1], l)
		for (var i = 0; i < r.length; ++i) print(r[i].join(''))
		print()
	}
}

File: sudoku/sudoku_v1.py
import sys, string

def sd_genmat():
	r, C = 0, []
	for i in xrange(9):
		for j in xrange(9):
			for k in xrange(9):
				C.append([ 9*i+j, i//3*27+j//3*9+k+81, 9*i+k+162, 9*j+k+243 ])
	R = [[] for c in xrange(324)]
	for r in xrange(729):
		for c2 in xrange(4):
			R[C[r][c2]].append(r)
	return R, C

def sd_update(R, C, sr, sc, r, v):
	m = 10
	m_c = 0
	for c2 in xrange(4): sc[C[r][c2]] += v<<7
	for c2 in xrange(4):
		c = C[r][c2]
		if v > 0:
			for r2 in xrange(9):
				rr = R[c][r2]
				sr[rr] += 1
				if sr[rr] == 1:
					for cc2 in xrange(4):
						cc = C[rr][cc2]
						sc[cc] -= 1
						if sc[cc] < m:
							m = sc[cc]
							m_c = cc
		else:
			for r2 in xrange(9):
				rr = R[c][r2]
				sr[rr] -= 1
				if sr[rr] == 0:
					p = C[rr]
					sc[p[0]] += 1; sc[p[1]] += 1; sc[p[2]] += 1; sc[p[3]] += 1;
	return m, m_c

def sd_solve(R, C, s):
	ret, out, hints = [], [], 0
	sr = [0] * 729
	sc = [9] * 324
	cr = [-1] * 81
	cc = [-1] * 81
	for i in xrange(81):
		if ord(s[i]) >= 49 and ord(s[i]) <= 57: a = ord(s[i]) - 49
		else: a = -1
		if a >= 0:
			sd_update(R, C, sr, sc, i * 9 + a, 1)
			hints += 1
		out.append(a + 1)
	i, m, d = 0, 10, 1
	while True:
		while i >= 0 and i < 81 - hints:
			if d == 1:
				if m > 1:
					for c in xrange(324):
						if sc[c] < m:
							m = sc[c]
							cc[i] = c
							if m < 2: break
				if m == 0 or m == 10:
					cr[i] = -1
					d = -1
					i -= 1
			c = cc[i]
			if d == -1 and cr[i] >= 0: sd_update(R, C, sr, sc, R[c][cr[i]], -1)
			r2_ = 9
			for r2 in xrange(cr[i] + 1, 9):
				if sr[R[c][r2]] == 0: r2_ = r2; break
			if r2_ < 9:
				m, cc[i+1] = sd_update(R, C, sr, sc, R[c][r2], 1)
				cr[i] = r2
				d = 1
				i += 1
			else:
				cr[i] = -1
				d = -1
				i -= 1
		if i < 0: break
		y = out[:81]
		for j in xrange(i):
			r = R[cc[j]][cr[j]]
			y[r//9] = r%9 + 1
		ret.append(y)
		i -= 1
		d = -1
	return ret

R, C = sd_genmat()
for line in sys.stdin:
	if len(line) >= 81:
		ret = sd_solve(R, C, line)
		for j in ret:
			print(''.join(map(str, j)))
		print('')
